File: src/main/java/com/catboyranch/roborancher/commands/CommandManager.java
package com.catboyranch.roborancher.commands;

import com.catboyranch.roborancher.utils.RoleType;
import com.catboyranch.roborancher.ServerConfig;
import com.catboyranch.roborancher.utils.ChannelUtils;
import com.catboyranch.roborancher.utils.CommandArgument;
import com.catboyranch.roborancher.utils.RoleUtils;
import net.dv8tion.jda.api.entities.ChannelType;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.emoji.Emoji;
import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import com.catboyranch.roborancher.RoboRancher;
import com.catboyranch.roborancher.Server;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;

public class CommandManager extends ListenerAdapter {
    private final RoboRancher rancher;
    public final HashMap<String, CommandBase> commands = new HashMap<>();

    public CommandManager(RoboRancher rancher) {
        this.rancher = rancher;
        rancher.getJDA().addEventListener(this);
    }

    @Override
    public void onMessageReceived(MessageReceivedEvent event) {
        //Don't reply to dm's or other bots
        if(event.getChannelType() == ChannelType.PRIVATE || event.getAuthor().isBot())
            return;

        Member member = event.getMember();
        if(member == null)
            return;
        Server server = rancher.getServer(event.getGuild().getId());
        ServerConfig cfg = server.getConfig();
        String msgContent = event.getMessage().getContentRaw();

        //Run filters
        if(cfg.isDeleteFilter() && !event.getGuildChannel().asTextChannel().isNSFW() && !RoleUtils.hasAdminPermission(member) && !RoleUtils.hasRole( member, cfg.getAdminRole())) {
            if(cfg.isFilterHard(msgContent) || cfg.isFilterSoft(msgContent))
                ChannelUtils.deleteMessage(event.getMessage());
        }

        String prefix = cfg.getCmdPrefix();
        if (msgContent.startsWith(prefix)) {
            msgContent = msgContent.replace(prefix, "");
            String[] argsRaw = msgContent.split(" ");

            String cmd = argsRaw[0];
            if (!commands.containsKey(cmd))
                return;

            argsRaw = Arrays.copyOfRange(argsRaw, 1, argsRaw.length);

            ArrayList<CommandArgument> argsList = new ArrayList<>();
            for(String arg : argsRaw) {
                if(!arg.equals(" ")) {
                    argsList.add(new CommandArgument(server, arg));
                }
            }
            CommandArgument[] args = new CommandArgument[argsList.size()];
            for(int i = 0; i < argsList.size(); i++) {
                args[i] = argsList.get(i);
            }

            boolean hasAdminPerm = RoleUtils.hasAdminPermission(member);
            CommandBase c = commands.get(cmd);
            boolean runCommand = false;

            if(!hasAdminPerm) {
                for (RoleType role : c.getAllowedRoles()) {
                    if (RoleUtils.hasRole(server, member, role)) {
                        runCommand = true;
                        break;
                    }
                }
            }

            if (runCommand || hasAdminPerm) {
                CommandResult result = new CommandResult() {
                    final Message msg = event.getMessage();

                    private void r(String emoji) {
                        msg.addReaction(Emoji.fromUnicode(emoji)).queue();
                    }

                    @Override
                    public void success() {
                        r("✅");
                    }

                    @Override
                    public void error(String message) {
                        r("❌");
                        msg.reply(message).queue();
                    }

                    @Override
                    public void successQuiet() { }
                };
                c.run(cmd, server, event, result, args);
            }
        }
    }

    public void registerCommand(CommandBase cmd) {
        commands.put(cmd.getName(), cmd);
    }

}

File: src/main/kotlin/com/catboyranch/roborancher/commands/CommandManager.kt
package com.catboyranch.roborancher.commands

import com.catboyranch.roborancher.ChannelUtils
import com.catboyranch.roborancher.RoboRancher
import com.catboyranch.roborancher.RoleUtils
import com.catboyranch.roborancher.Server
import net.dv8tion.jda.api.entities.ChannelType
import net.dv8tion.jda.api.entities.emoji.Emoji
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter
import java.util.*
import kotlin.collections.HashMap

class CommandManager(private val rancher: RoboRancher): ListenerAdapter() {
    private val commands = HashMap<String, CommandBase>()

    init {
        rancher.getJDA().addEventListener(this)
    }

    fun registerCommand(command: CommandBase) {
        commands[command.name] = command
    }

    override fun onMessageReceived(event: MessageReceivedEvent) {
        if(event.channelType == ChannelType.PRIVATE || event.author.isBot)
            return

        val member = event.member ?: return
        val server = rancher.getServer(event.guild.id) ?: return
        val cfg = server.config
        var msgContent = event.message.contentRaw

        if(cfg.getData().deleteFilter && !event.guildChannel.asTextChannel().isNSFW && !RoleUtils.hasAdminPermission(member) && !RoleUtils.hasRole(server, member, RoleUtils.RoleType.ADMIN)) {
            if(cfg.isFilterHard(msgContent) || cfg.isFilterSoft(msgContent))
                ChannelUtils.deleteMessage(event.message);
        }

        val prefix = cfg.getData().cmdPrefix
        if(!msgContent.startsWith(prefix))
            return

        msgContent = msgContent.replace(prefix, "");
        val argsRaw = msgContent.split(" ").toMutableList();

        val cmd = argsRaw.removeFirst()
        if (!commands.containsKey(cmd)) return

        val argsList = ArrayList<CommandArgument>()
        argsRaw.forEach {
            if(it != " ")
                argsList.add(CommandArgument(server, it))
        }
        val args = argsList.toTypedArray()
        val hasAdminPerm = RoleUtils.hasAdminPermission(member)
        val cmdBase = commands[cmd] ?: return
        var runCommand = false

        if(!hasAdminPerm) {
            cmdBase.allowedRoles.forEach {
                if(RoleUtils.hasRole(server, member, it))
                    runCommand = true
            }
        }

        if(runCommand || hasAdminPerm) {
            val result = object: CommandResult {
                val msg = event.message
                fun r(emoji: String) = msg.addReaction(Emoji.fromUnicode(emoji)).queue()

                override fun success() = r("✅")
                override fun successQuiet() { }
                override fun error(message: String) {
                    r("❌")
                    msg.reply(message).queue()
                }
            }
            cmdBase.run(cmd, server, event, result, *args)
        }
    }
}

abstract class CommandBase(private val rancher: RoboRancher) {
    abstract val name: String
    abstract val description: String
    abstract val allowedRoles: Array<RoleUtils.RoleType>

    abstract fun run(cmd: String, server: Server, event: MessageReceivedEvent, result: CommandResult, vararg args: CommandArgument)
}

interface CommandResult {
    fun success()
    fun successQuiet()
    fun error(message: String)
}